contrib/pg_tde/src/access/pg_tde_xlog_smgr.c

```diff
@@ -299,26 +336,64 @@ static ssize_t
 tdeheap_xlog_seg_write(int fd, const void *buf, size_t count, off_t offset,
 					   TimeLineID tli, XLogSegNo segno, int segSize)
 {
+	bool		lastKeyUsable;
+	bool		afterLastKey;
+#ifdef FRONTEND
+	bool		crashRecovery = false;
+#else
+	bool		crashRecovery = GetRecoveryState() == RECOVERY_STATE_CRASH;
+#endif
+
+	WalLocation loc = {.tli = tli};
+	WalLocation last_key_loc;
+
+	XLogSegNoOffsetToRecPtr(segno, offset, segSize, loc.lsn);
+
 	/*
 	 * Set the last (most recent) key's start LSN if not set.
 	 *
 	 * This func called with WALWriteLock held, so no need in any extra sync.
 	 */
-	if (EncryptionKey.type != WAL_KEY_TYPE_INVALID && TDEXLogGetEncKeyLsn() == 0)
-	{
-		WalLocation loc = {.tli = tli};
 
-		XLogSegNoOffsetToRecPtr(segno, offset, segSize, loc.lsn);
+	last_key_loc.lsn = TDEXLogGetEncKeyLsn();
+	pg_read_barrier();
+	last_key_loc.tli = TDEXLogGetEncKeyTli();
+
+	lastKeyUsable = (TDEXLogGetEncKeyLsn() != 0);
+	afterLastKey = wal_location_cmp(last_key_loc, loc) <= 0;
 
-		pg_tde_wal_last_key_set_location(loc);
-		EncryptionKey.wal_start = loc;
-		TDEXLogSetEncKeyLocation(EncryptionKey.wal_start);
+	if (EncryptionKey.type != WAL_KEY_TYPE_INVALID && !lastKeyUsable)
```

Comment:
Can't actually ever be invalid unless we have a bug? Maybe it should be an assert?